kratos/utilities/geometrical_sensitivity_utility.cpp

```diff
@@ -0,0 +1,282 @@
+
+    ///@}
+};
+
+///@} // Kratos Classes
+///@} // Adjoint Fluid Application group
+
+GeometricalSensitivityUtility::GeometricalSensitivityUtility(const JacobianType& rJ, const ShapeFunctionsLocalGradientType& rDN_De)
+: mpImpl(new Impl(rJ, rDN_De))
+{
+    KRATOS_TRY;
+    mpImpl->Initialize();
+    KRATOS_CATCH("");
+}
+
+GeometricalSensitivityUtility::~GeometricalSensitivityUtility() = default;
+
+void GeometricalSensitivityUtility::CalculateSensitivity(IndexType iNode, IndexType iCoord, double& rDetJ_Deriv, ShapeFunctionsGradientType& rDN_DX_Deriv) const
+{
+    KRATOS_TRY;
+    mpImpl->CalculateSensitivity(iNode, iCoord, rDetJ_Deriv, rDN_DX_Deriv);
+    KRATOS_CATCH("");
+}
+
+void GeometricalSensitivityUtility::Impl::Initialize()
+{
+    KRATOS_TRY;
+
+    KRATOS_ERROR_IF(mrJ.size1() != mrJ.size2()) << "Non-square Jacobian matrix." << std::endl;
+
+    KRATOS_ERROR_IF(mrJ.size2() != mrDN_De.size2())
+        << "Jacobian local-coordinates size (" << mrJ.size2()
+        << ") != shape function local-coordinates size(" << mrDN_De.size2()
+        << ")." << std::endl;
+
+    mCofactorJ = CalculateCofactorMatrix(mrJ);
+    mDetJ = CalculateDeterminant(mrJ);
+
+    KRATOS_CATCH("");
+}
+
+void GeometricalSensitivityUtility::Impl::CalculateSensitivity(IndexType iNode, IndexType iCoord, double& rDetJ_Deriv, ShapeFunctionsGradientType& rDN_DX_Deriv) const
+{
+    KRATOS_TRY;
+
+    rDetJ_Deriv = CalculateDeterminantOfJacobianSensitivity(iNode, iCoord);
+
+    MatrixType cofactorJ_deriv = CalculateCofactorOfJacobianSensitivity(iNode, iCoord);
+    if (rDN_DX_Deriv.size1() != mrDN_De.size1() || rDN_DX_Deriv.size2() != mCofactorJ.size1())
+        rDN_DX_Deriv.resize(mrDN_De.size1(), mCofactorJ.size1());
+    noalias(rDN_DX_Deriv) = (1.0 / mDetJ) * prod(mrDN_De, trans(cofactorJ_deriv));
+    rDN_DX_Deriv += -(rDetJ_Deriv / (mDetJ * mDetJ)) * prod(mrDN_De, trans(mCofactorJ));
+
+    KRATOS_CATCH("");
+}
+
+template<class TMatrixType>
+double GeometricalSensitivityUtility::Impl::CalculateDeterminant(const TMatrixType& rMat) const
+{
+    KRATOS_TRY;
+
+    double det;
+    if (rMat.size1() == 1)
+        det = rMat(0, 0);
+    else if (rMat.size1() == 2)
+        det = rMat(0, 0) * rMat(1, 1) - rMat(1, 0) * rMat(0, 1);
+    else if (rMat.size1() == 3)
+        det = rMat(0, 0) * rMat(1, 1) * rMat(2, 2) +
+              rMat(1, 0) * rMat(2, 1) * rMat(0, 2) +
+              rMat(0, 1) * rMat(1, 2) * rMat(2, 0) -
+              rMat(2, 0) * rMat(1, 1) * rMat(0, 2) -
+              rMat(2, 1) * rMat(1, 2) * rMat(0, 0) -
+              rMat(1, 0) * rMat(0, 1) * rMat(2, 2);
+    else
+        KRATOS_ERROR << "Matrix dimension = " << rMat.size1()
+                     << " is not supported." << std::endl;
+
+    return det;
+    KRATOS_CATCH("");
+}
+
+template <class TMatrixType>
+double GeometricalSensitivityUtility::Impl::CalculateCofactor(const TMatrixType& rMat, IndexType i, IndexType j) const
```

Comment:
I'd also suggest placing this on a more generic utility (math utils or otherwise), since it is a generic math tool that may be used beyond this class.